### ISSUE TITLE
Legacy API Deprecation

### ISSUE THREAD

**Issue by euantorano** [firebase/firebase-admin-dotnet#349]
We've received an email yesterday regarding deprecation of the legacy API for FCM. It links to [an FAQ](https://firebase.google.com/support/faq#fcm-23-deprecation) entry, which says to migrate usages of `SendMulticastAsync` to `SendEachForMulticastAsync`.

I see this method was added in commit 74bd9e50, but there has not been a release tagged since 24/03/2022.

Is there any chance a new release might be tagged and pushed to NuGet at some point soon (preferably before the API cutoff on 20/06/2024...)?

Personally, I wouldn't have published an FAQ or sent out a deprecation warning email until all of my SDKs were updated, but that's just me...

**Comment by euantorano** [comment]
I note that this seems to affect every version of the Admin SDK, including:

- .NET (this issue)
- Go
- Java (functions implemented in https://github.com/firebase/firebase-admin-java/commit/d1baee366cebf121aa6d5a62593c08cc26ce97a8 but no release tagged)
- Python (functions implemented in https://github.com/firebase/firebase-admin-python/commit/4323ed88cef168d66844add3aeaf3b7d19b46be6 but no release tagged)

In fact, the only version that seems OK is Node (functions implemented in https://github.com/firebase/firebase-admin-node/commit/0da72ef60535b6a9d0267e3843f45512718b8e74 and seemingly included in release 11.9.0)

**Comment by fu-mybeepr** [comment]
I was wondering the same. The [FAQ page](https://firebase.google.com/support/faq#fcm-depr-admin) specifically says `>= 2.4.0` for .Net, but apparently 2.4.0 is not released on either GitHub or Nuget.

**Comment by tonsnoei** [comment]
In addition, I got the following advice of the Firebase guys for replacing the batch/multicast message: `"you should consider using HTTP/2 to send multiple requests over the same HTTP connection to the HTTP v1 API"`. I think, looking at the code, HTTP/2 is not supported yet. Are you planning that in the near future?

For completeness I added my question to the Firebase guys and the answer I got:

**Question**

> Subject: Multicast messaging removed?
> 
> I got a mail that batch messaging will be removed from the FCM API. I was wondering, if both multicast and batch-messages will be removed. Why removing this efficient way of communicating with the platform? For my app I send the same push message to a lot of users (sometimes 1000 to 2000) depending on their location. Currently I can do this with 4 calls (4 times 500 messages). In the future I have to perform 1000 to 2000 calls...
> 
> Just wondering... I hope to hear from you and thanks for taking the time to take a look at this request. Have a nice day.

**Answer**

> I understand that you’ve received an MSA informing you that some FCM APIs are being deprecated and will be decommissioned by June 2024 recently. This [section](https://firebase.google.com/support/faq#fcm-23-deprecation) of our Frequently Asked Questions page highlights the specific steps and processes that many developers such as yourself may want to follow to ensure that things work as expected after June 2024, so we usually recommend customers with similar inquiries start from there. I’d also suggest reading this [guide](https://firebase.google.com/docs/cloud-messaging/migrate-v1) about migrating to the HTTPv1 API. Of course, I’ll do my best to address your questions below.
> 
> This feature, called "multicast" in legacy HTTP APIs, is not supported by the HTTP v1 API. For use cases where end-to-end latency is critical, or where total fanout size is small (fewer than 1 million), Google recommends the HTTP V1 API. The HTTP V1 API over HTTP/2 can achieve up to ~1/3 of total message throughput rate compared to legacy multicast when sending on a single connection.
> 
> For use cases where throughput and egress bandwidth are prioritized or where total fanout size is large (greater than 1 million), Google recommends topic messaging. Topics are the most widely supported mechanism for sending to multiple devices and are usually able to handle most use cases in a meaningful way. Plus, they’re generally simple to understand and use, and developers can create an unlimited number of topics. So, I’m pretty confident that you will be able to translate your implementation to topics with a bit of experimentation. 
> 
> Lastly, as an alternative to the legacy batch send method, you may use the [standard HTTP v1 API send method](https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages/send), implementing your own batch send by iterating through the list of recipients and sending to each recipient's token. The FCM batch send API uses the same [message format](https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages#Message) and authentication mechanism as the HTTP v1 API. However, it uses a different endpoint. If you want to improve efficiency, you should consider using HTTP/2 to send multiple requests over the same HTTP connection to the HTTP v1 API.
> 
> Hope this helps point you in the right direction, or at least gives you some options to consider. Don’t hesitate to reconnect should you have any additional questions, concerns, clarifications, or comments after considering the above and we can dive deeper.

**Comment by tonsnoei** [comment]
I saw the changes to `2.4.0`, however the suggested usage of HTTP/2 for the HttpClient is not implemented yet. Right? 

I am not sure how the HttpClient behaves. Does it automatically use HTTP/2 in some situations? From what I understand you have to set the handler of the `HttpClient` the following way:

```
DefaultRequestVersion = new Version(2, 0); // Use HTTP/2
```

**Comment by ewfian** [comment]
> Additional to the above questions and important in relation to the performance of the library with this new (batch/multicast replacing) functionality:
> 
> I saw that in `2.4.0` the suggested usage of HTTP/2 for the HttpClient is not implemented yet. Right?
> 
> I am not sure how the HttpClient behaves. Does it automatically use HTTP/2 in some situations? From what I understand you have to set the handler of the `HttpClient` explicitly in the following way:
> 
> ```
> DefaultRequestVersion = new Version(2, 0); // Use HTTP/2
> ```
> 
> Just wondering...
> 
> Furthermore, keep up the good work!


Same question here.
From this doc, the default version is 1.1 on .NET 6
> In .NET Core, the default message version differs based on the version you're using. .NET Core 2.1 changed the default value from 1.1 to 2.0. In .NET Core 3.0, the default value was reverted back to **1.1.**
> https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httprequestmessage.version?view=net-6.0#property-value


FirebaseAdmin supports `netstandard2.0`, while `netstandard2.0` compatible with  .NET Core 2.1 and 2.2
https://github.com/firebase/firebase-admin-dotnet/blob/bc7bc090a0636b4c7487dd530511139ec60b167a/FirebaseAdmin/FirebaseAdmin/FirebaseAdmin.csproj#L5C30-L5C30
https://learn.microsoft.com/en-us/dotnet/standard/net-standard?tabs=net-standard-2-0

According this comment, the default behavior will vary between different .net versions
> The HTTP message version. The default value is 1.1, unless you're targeting .NET Core 2.1 or 2.2. In that case, the default value is 2.0.

**Comment by euantorano** [comment]
Now that 2.4.0 has been released, I've updated my project to it.

It looks like `SendMulticastAsync` hasn't been marked as obsolete/deprecated as the [`ObsoleteAttribute`](https://learn.microsoft.com/en-us/dotnet/api/system.obsoleteattribute?view=net-7.0) is commented out: https://github.com/firebase/firebase-admin-dotnet/blob/bc7bc090a0636b4c7487dd530511139ec60b167a/FirebaseAdmin/FirebaseAdmin/Messaging/FirebaseMessaging.cs#L422

It seems the attribute was added but commented out in https://github.com/firebase/firebase-admin-dotnet/commit/74bd9e5038aa49562a093fa72c80c252384e85c8 - this is fairly easy to miss as Visual Studio by default continues the block comment when you enter a new line at the end.

**Comment by lahirumaramba** [comment]
Hey folks, you are right there was a delay on the .NET release and the new batch send APIs are now available in `v2.4.0`
https://firebase.google.com/support/release-notes/admin/dotnet#version_240_-_12_july_2023

Thank you for flagging the issue with the `Obsolete` tag! @Doris-Ge could you take a look at the [ObsoleteAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.obsoleteattribute?view=net-7.0) when you get a chance, please? Thanks!

**Comment by Doris-Ge** [comment]
> Thank you for flagging the issue with the `Obsolete` tag! @Doris-Ge could you take a look at the [ObsoleteAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.obsoleteattribute?view=net-7.0) when you get a chance, please? Thanks!

Sure. I think I misunderstood the usage of `Obsolete`. 

Thanks for flagging the issue! Will fix it.

**Comment by danports** [comment]
I upgraded a project to v2.4.0 and replaced usages of `SendMulticastAsync` with `SendEachForMulticastAsync` and had to revert the upgrade a few days later because `SendEachForMulticastAsync` did not deliver notifications to devices as reliably as `SendMulticastAsync` - some notifications seemed to be lost. Has anyone else that upgraded ran into this? I haven't yet dug into where in the stack the notifications were being lost, and it's possible it was a fluke.

**Comment by R4DIC4L** [comment]
Is this related to the legacy Firebase Cloud Messaging (FCM) APIs which will be discontinued on June 20, 2024?

Does `SendMulticastAsync` use Batch send API?

> On June 20, 2024, we’re reducing the number of Firebase Cloud Messaging (FCM) legacy register APIs and legacy send APIs that provide similar functionality. This step will allow us to provide you with a more consistent experience and align with Google security standards to improve security, reliability and performance.
> 
> Because of these API decommissions, some already-deprecated SDKs and features will stop working after June 20, 2024.
> 
> Please consult the tables below to find which Firebase Cloud Messaging (FCM) APIs and corresponding services/SDKs/features will be discontinued and replaced with new alternatives.

![image](https://github.com/firebase/firebase-admin-dotnet/assets/7274664/9b8a1aeb-949e-42a5-8103-a2a36b47c3dc)

**Comment by R4DIC4L** [comment]
> I upgraded a project to v2.4.0 and replaced usages of `SendMulticastAsync` with `SendEachForMulticastAsync` and had to revert the upgrade a few days later because `SendEachForMulticastAsync` did not deliver notifications to devices as reliably as `SendMulticastAsync` - some notifications seemed to be lost. Has anyone else that upgraded ran into this? I haven't yet dug into where in the stack the notifications were being lost, and it's possible it was a fluke.

I am now performing the same upgrade to v2.4.0, but need more time to test the new `SendEachForMulticastAsync` method for notifications. Any particular use case when this happens? Is it happening for a high volume of notifications?

**Comment by Doris-Ge** [comment]
> Is this related to the legacy Firebase Cloud Messaging (FCM) APIs which will be discontinued on June 20, 2024?
Does SendMulticastAsync use Batch send API?

Yes and yes.

**Comment by hanselsen** [comment]
Hi all. Since this issue is still open, I wish to inquire about the status of it.  
Reading back in this thread, it should be fixed when people upgrade to 2.4.0?

For example, I am using `SendAllAsync` and `SendAsync`. Do I need to do anything else?

**Comment by EricaLi123** [comment]
> Hi all. Since this issue is still open, I wish to inquire about the status of it. Reading back in this thread, it should be fixed when people upgrade to 2.4.0?
> 
> For example, I am using `SendAllAsync` and `SendAsync`. Do I need to do anything else?

You just need to replace ```SendAllAsync``` with ```SendEachAsync```.
[https://firebase.google.com/support/release-notes/admin/dotnet#cloud-messaging](url)

**Comment by alex-radch** [comment]
There are ~2 months left before the completion of firebase legacy methods, is there any news? Our company is waiting for a solution to this problem. Is there really only a manual way left for us, like presented above?

**Comment by tonsnoei** [comment]
> There are ~2 months left before the completion of firebase legacy methods, is there any news? Our company is waiting for a solution to this problem. Is there really only a manual way left for us, like presented above?

As far as I know, just replace the `SendAllAsync` with `SendEachAsync` as @EricaLi123 mentioned. We did this with our company a few months ago. It works well.

**Comment by Igouist** [comment]
> > Is this related to the legacy Firebase Cloud Messaging (FCM) APIs which will be discontinued on June 20, 2024?這與將於 2024 年 6 月 20 日停用的舊版 Firebase Cloud Messaging (FCM) API 有關嗎？
> > Does SendMulticastAsync use Batch send API?SendMulticastAsync 是否使用批量發送 API？
> 
> Yes and yes. 是的，是的。

I just want to confirm that we are currently using `SendMulticastAsync` . 

Should we switch to `SendEachForMulticastAsync` by June 20th, or am I misunderstanding something? 

Thank you QQ

**Comment by tonsnoei** [comment]
> > > Is this related to the legacy Firebase Cloud Messaging (FCM) APIs which will be discontinued on June 20, 2024?這與將於 2024 年 6 月 20 日停用的舊版 Firebase Cloud Messaging (FCM) API 有關嗎？
> > > Does SendMulticastAsync use Batch send API?SendMulticastAsync 是否使用批量發送 API？
> > 
> > 
> > Yes and yes. 是的，是的。
> 
> I just want to confirm that we are currently using `SendMulticastAsync` .
> 
> Should we switch to `SendEachForMulticastAsync` by June 20th, or am I misunderstanding something?
> 
> Thank you QQ
Yes, you should start using `SendEachForMulticastAsync` instead of `SendMulticastAsync`

**Comment by tonsnoei** [comment]
> > > Is this related to the legacy Firebase Cloud Messaging (FCM) APIs which will be discontinued on June 20, 2024?這與將於 2024 年 6 月 20 日停用的舊版 Firebase Cloud Messaging (FCM) API 有關嗎？
> > > Does SendMulticastAsync use Batch send API?SendMulticastAsync 是否使用批量發送 API？
> > 
> > 
> > Yes and yes. 是的，是的。
> 
> I just want to confirm that we are currently using `SendMulticastAsync` .
> 
> Should we switch to `SendEachForMulticastAsync` by June 20th, or am I misunderstanding something?
> 
> Thank you QQ


Yes! Use `SendEachForMulticastAsync` now instead of `SendMulticastAsync`

**Comment by Sankarvjss78** [comment]
**Step 1:**

go to https://developers.google.com/oauthplayground/

1.Select the autorize api and after fill the **`input your own scopes (email, https://www.googleapis.com/auth/firebase.messaging)`** once it is complete after displayed the client id and client secret key for refresh_token.


**Step 2:**
POST  => https://developers.google.com/oauthplayground/refreshAccessToken

{"token_uri":"https://oauth2.googleapis.com/token","refresh_token":"1//04BZzVDcaVl4vCgYIARAAGAQSNwF-L9IrPUzn_0****"}  in body


**Step 3:**

POST  =>> https://fcm.googleapis.com/v1/projects/chat-2dc2d/messages:send

headers: Bearer Y29**********.......

{
    "message": {
        "token": "*********",
        "notification": {
            "title": "Sample",
            "body": "Sample"
        }
    }
}  in body

**Comment by uchida-nunet** [comment]
### Performance of the SendEachForMulticastAsync methods is too slow

After updating to v3.0.1, I am facing performance issues.

I am passing 500 tokens to the new SendEachForMulticastAsync for sending, but it takes several times (up to 20 times) longer compared to the old SendMulticastAsync.

Is there a way to resolve this? Also, from what I can see in this thread, it seems that sending via HTTP 2.0 may improve performance. Can this be applied to my project (.Net Framework 4.7.2)?

**Comment by nbevans** [comment]
Indeed, performance is diabolical with this new SendEachForMulticastAsync method

The HttpClient it uses is not configured to use HTTP 2.0. For .NET Framework, it needs to use the System.Net.Http.WinHttpHandler nuget which is what provides HTTP 2.0 support on NetFx.

**Comment by uchida-nunet** [comment]
@nbevans 

Thank you for the information. I'll look into configuring the HttpClient with HTTP 2.0 support using the `System.Net.Http.WinHttpHandler` for .NET Framework as suggested. I appreciate your help!

**Comment by uchida-nunet** [comment]
Thank you! I’ll try a few different approaches and see what works.
